File: day03.js
const LINE_RE = /^#(\d+) @ (\d+),(\d+): (\d+)x(\d+)$/;

const forEachLocation = (startX, startY, w, h, callback) => {
  for (let x = startX; x < startX + w; x++) {
    for (let y = startY; y < startY + h; y++) {
      callback([x, y].join(","));
    }
  }
};

export function part1(lines) {
  const covered = {};

  for (const line of lines) {
    const match = LINE_RE.exec(line);
    const [startX, startY, w, h] = match.slice(2).map(Number);
    forEachLocation(startX, startY, w, h, (key) => {
      covered[key] = (covered[key] || 0) + 1;
    });
  }

  return Object.entries(covered).reduce(
    (count, [_, v]) => (v > 1 ? count + 1 : count),
    0
  );
}

export function part2(lines) {
  const claims = [];
  const claimLocations = {};

  for (const line of lines) {
    const match = LINE_RE.exec(line);
    const [id, startX, startY, w, h] = match.slice(1).map(Number);
    claims.push([id, startX, startY, w, h]);
    forEachLocation(startX, startY, w, h, (key) => {
      claimLocations[key] = (claimLocations[key] || []).concat([id]);
    });
  }

  for (const [id, startX, startY, w, h] of claims) {
    let unique = true;
    forEachLocation(startX, startY, w, h, (key) => {
      unique =
        unique &&
        claimLocations[key].length === 1 &&
        claimLocations[key][0] === id;
    });
    if (unique) {
      return id;
    }
  }
}

File: day03.py
import re
from collections import Counter, defaultdict

from base import BaseSolution


LINE_RE = re.compile(r"^#(\d+) @ (\d+),(\d+): (\d+)x(\d+)$")


class Solution(BaseSolution):
    def load_data(self, input):
        return [
            tuple(int(i) for i in LINE_RE.match(line).groups())
            for line in input.splitlines()
        ]

    def part1(self, data):
        counts = Counter(
            coords for _, *area in data for coords in self.get_all_coords(*area)
        )
        return sum(count > 1 for count in counts.values())

    def part2(self, entries):
        claims = defaultdict(set)
        for claim_id, *area in entries:
            for coords in self.get_all_coords(*area):
                claims[coords].add(claim_id)
        for claim_id, *area in entries:
            if all(
                claims[coords] == {claim_id} for coords in self.get_all_coords(*area)
            ):
                return claim_id

    def get_all_coords(self, start_x, start_y, w, h):
        return (
            (x, y)
            for x in range(start_x, start_x + w)
            for y in range(start_y, start_y + h)
        )
